test/fixtures/hdfs-fixture/src/main/java/org/elasticsearch/test/fixtures/hdfs/HdfsFixture.java

```diff
@@ -0,0 +1,444 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.test.fixtures.hdfs;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.CommonConfigurationKeysPublic;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.FileUtil;
+import org.apache.hadoop.fs.permission.AclEntry;
+import org.apache.hadoop.fs.permission.AclEntryType;
+import org.apache.hadoop.fs.permission.FsAction;
+import org.apache.hadoop.ha.BadFencingConfigurationException;
+import org.apache.hadoop.ha.HAServiceProtocol;
+import org.apache.hadoop.ha.HAServiceTarget;
+import org.apache.hadoop.ha.NodeFencer;
+import org.apache.hadoop.ha.ZKFCProtocol;
+import org.apache.hadoop.ha.protocolPB.HAServiceProtocolClientSideTranslatorPB;
+import org.apache.hadoop.hdfs.DFSConfigKeys;
+import org.apache.hadoop.hdfs.MiniDFSCluster;
+import org.apache.hadoop.hdfs.MiniDFSNNTopology;
+import org.apache.hadoop.hdfs.server.namenode.ha.HATestUtil;
+import org.apache.hadoop.hdfs.tools.DFSHAAdmin;
+import org.apache.hadoop.security.SecurityUtil;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.junit.Assume;
+import org.junit.rules.ExternalResource;
+import org.junit.rules.TemporaryFolder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.security.AccessController;
+import java.security.PrivilegedActionException;
+import java.security.PrivilegedExceptionAction;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Supplier;
+
+public class HdfsFixture extends ExternalResource {
+
+    private TemporaryFolder temporaryFolder = new TemporaryFolder();
+    private MiniDFSCluster dfs;
+    private String haNameService;
+    private Supplier<String> principalConfig = null;
+    private Supplier<Path> krb5Conf = null;
+    private Supplier<Path> keytab = null;
```

Comment:
Could we simplify this to just accept a `String`. This would avoid the need to create a temporary file in `Krb5kDcContainer` as well.

Reply:
Not sure I get what you mean. The `keytab` file within the container is generated at startup time, so _after_ the hdfsfixture constructor, We don't have the content of that file available earlier. that's why we resolve this only when we start the hdfs cluster as the krb5kdc container has been started and key tab is available.

Reply:
Sure, what I mean is we only need the _contents_ of the file, we don't actually need a path too a file on disk. And in the krb fixture we have to read the file out of the container, and then write it to disk, then return the file path. We could avoid that intermediary step of creating a temporary file on disk and just return the content of the file as a string.

Reply:
Or maybe I'm mistaken and the fixture does require this configuration to provided in a the form of a file location, in which case nevermind.